File: legacy/packages/core/src/actions/query/read.ts
import { z } from "zod";
import { ActionHandler, resolveExecutor } from "../../types.js";

export const ReadSchema = z.object({
    action: z.literal("read").describe("Read action - execute SELECT queries"),
    sql: z.string().describe("SQL SELECT query to execute"),
    params: z.array(z.unknown()).optional().describe("Parameterized query values (use $1, $2, etc. in SQL)"),
    session_id: z.string().optional().describe("Session ID returned by pg_tx 'begin'. Use to read uncommitted data within a transaction."),
    options: z.object({
        timeout_ms: z.number().optional().describe("Query timeout in milliseconds"),
    }).optional().describe("Query execution options"),
});

export const readHandler: ActionHandler<typeof ReadSchema> = {
    schema: ReadSchema,
    handler: async (params, context) => {
        const executor = resolveExecutor(context, params.session_id);
        const options: any = {};
        if (params.options?.timeout_ms !== undefined) {
            options.timeout_ms = params.options.timeout_ms;
        }
        return await executor.execute(params.sql, params.params, options);
    },
};

File: coldquery/actions/query/read.py
from typing import Any, Dict, List, Optional
from fastmcp.exceptions import ToolError

from coldquery.core.context import ActionContext, resolve_executor
from coldquery.core.executor import QueryResult
from coldquery.middleware.session_echo import enrich_response


async def read_handler(params: Dict[str, Any], context: ActionContext) -> str:
    """Handles the 'read' action to execute SELECT queries."""
    sql: Optional[str] = params.get("sql")
    query_params: Optional[List[Any]] = params.get("params")
    session_id: Optional[str] = params.get("session_id")

    if not sql:
        raise ToolError("The 'sql' parameter is required for the 'read' action.")

    executor = await resolve_executor(context, session_id)
    result: QueryResult = await executor.execute(sql, query_params)

    return enrich_response(
        result.to_dict(), session_id, context.session_manager
    )
